src/timed.rs

```diff
@@ -0,0 +1,220 @@
+        self.send(TimedMessage::Recurring { factory: Box::new(factory), fire_at, interval })
+    }
+}
+
+/// A [`Context`] variant available to actors wrapped by the [`Timed`] actor wrapper.
+pub struct TimedContext<M> {
+    pub system_handle: SystemHandle,
+    pub myself: Recipient<TimedMessage<M>>,
+}
+
+impl<M> TimedContext<M> {
+    fn from_context(context: &Context<TimedMessage<M>>) -> Self {
+        TimedContext {
+            system_handle: context.system_handle.clone(),
+            myself: context.myself.clone(),
+        }
+    }
+}
+
+/// A wrapper around actors to add ability to receive delayed and recurring messages.
+/// See [module documentation](self) for a complete recipe.
+pub struct Timed<A: Actor> {
+    inner: A,
+    queue: BinaryHeap<QueueItem<A::Message>>,
+}
+
+impl<M: Send + 'static, A: Actor<Context = TimedContext<M>, Message = M>> Timed<A> {
+    pub fn new(inner: A) -> Self {
+        Self { inner, queue: Default::default() }
+    }
+
+    fn schedule_timeout(&self, context: &mut <Self as Actor>::Context) {
+        // Schedule next timeout if the queue is not empty.
+        context.set_deadline(self.queue.peek().map(|earliest| earliest.fire_at));
+    }
+}
+
+impl<M: Send + 'static, A: Actor<Context = TimedContext<M>, Message = M>> Actor for Timed<A> {
+    type Context = Context<Self::Message>;
+    type Error = A::Error;
+    type Message = TimedMessage<M>;
+
+    fn handle(
+        &mut self,
+        context: &mut Self::Context,
+        timed_message: Self::Message,
+    ) -> Result<(), Self::Error> {
+        let item = match timed_message {
+            TimedMessage::Instant { message } => {
+                return self.inner.handle(&mut TimedContext::from_context(context), message);
+            },
+            TimedMessage::Delayed { message, fire_at } => {
+                QueueItem { fire_at, payload: Payload::Delayed { message } }
+            },
+            TimedMessage::Recurring { factory, fire_at, interval } => {
+                QueueItem { fire_at, payload: Payload::Recurring { factory, interval } }
+            },
+        };
+
+        self.queue.push(item);
+        self.schedule_timeout(context);
+        Ok(())
+    }
+
+    fn name() -> &'static str {
+        A::name()
+    }
+
+    fn started(&mut self, context: &mut Self::Context) {
+        self.inner.started(&mut TimedContext::from_context(context))
+    }
+
+    fn stopped(&mut self, context: &mut Self::Context) {
+        self.inner.stopped(&mut TimedContext::from_context(context))
+    }
+
+    fn deadline_passed(
+        &mut self,
+        context: &mut Self::Context,
+        _deadline: Instant,
+    ) -> Result<(), Self::Error> {
+        // Handle all messages that should have been handled by now.
+        let now = Instant::now();
+        while self.queue.peek().map(|m| m.fire_at <= now).unwrap_or(false) {
+            let item = self.queue.pop().expect("heap is non-empty, we have just peeked");
```

Comment:
Not a comment on your code, but it'd be nice if they also had a [`drain_filter()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain_filter) function for `BinaryHeap`.

Reply:
It would! I don't like using `expect()` for something that could be proved infallible at compile time.